src/dialer/token-holder.js

```diff
@@ -0,0 +1,63 @@
+'use strict'
+
+/**
+ * @class TokenHolder
+ * @example
+ * const th = new TokenHolder(tokens, dialer.releaseToken)
+ * for (const action of actions) {
+ *   const token = await th.getToken()
+ *   action(token).then(() => th.releaseToken(token))
+ * }
+ *
+ * await th.drain()
+ */
+class TokenHolder {
+  /**
+   * @param {Array<*>} tokens Tokens to track
+   * @param {function(*)} release Called when releasing control of the tokens
+   */
+  constructor (tokens, release) {
+    this.originalTokens = tokens
+    this.tokens = [...tokens]
+    this._release = release
+  }
+
+  /**
+   * Resolves a token once once is available. Once the token is no
+   * longer needed it MUST be release with `releaseToken()`.
```

Comment:
Imagine this scenario: We ask for 3 tokens (and we have 5 multiaddr to try), but we only got 1 token.
Can we try to get another one during the dial process? We can make this improvement later on, but everytime we block in this function, we could try to ask for more first, in the event that we did not receive all of them in the first ask

Reply:
Yeah, I think there is potential here to clean a lot of the token handling up. So the dialer might give you a token generator, and it will eventually put the `n` tokens you asked for into the generator. The dial process can then pull from that generator and potentially push into a local generator. The local generator is actually a token cycler. So when a multiaddr dial is done but there are more to do, it pushes the token back into the cycler for other dials. Once there is no more dials to perform for a token, it's pushed into the dialer token pool. Psuedo code:

```js
pipe(
  dialer.getTokens(3),
  dialRequestTokenCycler(),  // Reuse tokens until they are not needed, then yield them
  dialer.consumeTokens()
)
```

Reply:
this would be super cool